mini/einstein_test.go

```diff
@@ -0,0 +1,194 @@
+package mini
+
+import (
+	"testing"
+	"time"
+
+	"github.com/awalterschulze/gominikanren/micro"
+	"github.com/awalterschulze/gominikanren/sexpr/ast"
+)
+
+/*
+Let us assume that there are five houses of different colors next to each other on the same road.
+In each house lives a man of a different nationality.
+Every man has his favorite drink, his favorite brand of cigarettes, and keeps pets of a particular kind.
+
+The Englishman lives in the red house.
+The Swede keeps dogs.
+The Dane drinks tea.
+The green house is just to the left of the white one.
+The owner of the green house drinks coffee.
+The Pall Mall smoker keeps birds.
+The owner of the yellow house smokes Dunhills.
+The man in the center house drinks milk.
+The Norwegian lives in the first house.
+The Blend smoker has a neighbor who keeps cats.
+The man who smokes Blue Masters drinks bier.
+The man who keeps horses lives next to the Dunhill smoker.
+The German smokes Prince.
+The Norwegian lives next to the blue house.
+The Blend smoker has a neighbor who drinks water.
+
+The question to be answered is: Who keeps fish?
+Answer adapted from https://gist.github.com/jgilchrist/1b85b04f4395057972dd
+*/
+
+// takes almost a minute (52s) to run atm on my machine with MemberO
+// takes about 5s to run with MemberO unrolled, _without any concurrency support_
```

Comment:
For reference, the minikanren-in-scheme version runs in 0.35s or so, without unrolling.
We have some optimisation to do :)

Reply:
One more note: if I change the substitution map from string->sexpr to int64->sexpr, we shave off another 10% running time or so. It also brings us more inline with existing ukanren/minikanren implementations.
I think that is one more thing we need can do (separate PR), but I run into issues with the parser doing it.
In order to run it locally I had to delete all other tests, which is not ideal ;)

Reply:
If you go that route, go also try uint64.  I recall it being faster for map lookup, but that was 6 years ago.

Although a 10% shaving is far from the two orders of magnitude that we need.  So might be worth looking for some bigger wins first, but that is your choice.  I just find the strings easier to read.

Reply:
tbh the main argument for doing it early is that we can compare with other implementations.
This variable naming and conflation of concepts with the parser has confused a bunch of times already.

I am comparing with fast-minikanren btw, so that is optimised in a bunch of ways. 
Not sure how naive minikanren-in-chezscheme would perform.
I've been digging deep into the fast-minikanren implementation and almost have a grasp of all the moving parts;
there's a lot of things we can do with checking the scope of vars. But again, we need vars to be cleaned up first.

I kinda need you for the parser bit, since I'm not sure on the details there and what might break.
Haven't ever _used_ the parser either, just writting gominikanren funcs by hand in Go :)

Reply:
Cool sounds good.  Lets also discuss the parser on Tuesday :)